packages/tools/src/lib/build-cache/fetchCachedBuild.ts

```diff
@@ -61,32 +66,103 @@ Proceeding with local build.`);
+    loader
+  );
+  await extractArtifactTarballIfNeeded(localArtifactPath);
+  const binaryPath = getLocalBinaryPath(localArtifactPath);
   if (!binaryPath) {
     loader.stop(`No binary found for "${artifactName}".`);
     return null;
   }
-
   loader.stop(
     `Downloaded cached build: ${color.cyan(path.relative(root, binaryPath))}.`
   );
 
   return {
-    name: fetchedBuild.name,
-    artifactPath: fetchedBuild.path,
+    name: artifactName,
+    artifactPath: localArtifactPath,
     binaryPath,
   };
 }
+
+async function handleDownloadResponse(
+  response: Response,
+  localArtifactPath: string,
+  name: string,
+  loader: ReturnType<typeof spinner>
+) {
+  try {
+    fs.mkdirSync(localArtifactPath, { recursive: true });
+    if (!response.ok || !response.body) {
+      throw new Error(`Failed to download artifact: ${response.statusText}`);
+    }
+    let responseData = response;
+    const contentLength = response.headers.get('content-length');
+
+    if (contentLength) {
+      const totalBytes = parseInt(contentLength, 10);
+      const totalMB = (totalBytes / 1024 / 1024).toFixed(2);
+      let downloadedBytes = 0;
+
+      const reader = response.body.getReader();
+      const stream = new ReadableStream({
+        async start(controller) {
+          while (true) {
+            const { done, value } = await reader.read();
+            if (done) {
+              break;
+            }
+            downloadedBytes += value.length;
+            const progress = ((downloadedBytes / totalBytes) * 100).toFixed(0);
+            loader?.message(
+              `Downloading cached build from ${name} (${progress}% of ${totalMB} MB)`
+            );
+            controller.enqueue(value);
+          }
+          controller.close();
+        },
+      });
+      responseData = new Response(stream);
+    }
+
+    const zipPath = localArtifactPath + '.zip';
+    const buffer = await responseData.arrayBuffer();
+    fs.writeFileSync(zipPath, new Uint8Array(buffer));
+    unzipFile(zipPath, localArtifactPath);
+    fs.unlinkSync(zipPath);
+  } catch (error) {
+    loader?.stop(`Failed: Downloading cached build from ${name}`);
+    throw new RnefError(`Unexpected error`, { cause: error });
+  }
+}
+
+function unzipFile(zipPath: string, targetPath: string): void {
+  const zip = new AdmZip(zipPath);
+  zip.extractAllTo(targetPath, true);
+}
+
+async function extractArtifactTarballIfNeeded(artifactPath: string) {
+  const tarPath = path.join(artifactPath, 'app.tar.gz');
+
+  // If the tarball is not found, it means the artifact is already unpacked.
+  if (!fs.existsSync(tarPath)) {
+    return;
+  }
+
+  // iOS simulator build artifact (*.app directory) is packed in .tar.gz file to
+  // preserve execute file permission.
+  // See: https://github.com/actions/upload-artifact?tab=readme-ov-file#permission-loss
+  await tar.extract({
+    file: tarPath,
+    cwd: artifactPath,
+    gzip: true,
+  });
+  fs.unlinkSync(tarPath);
+}
```

Comment:
probably it'll be fine, but we're coupling github-specific things to the function that will be used for all the possible remote cache providers.

Reply:
yeah, I'm pretty sure I'll need to rewrite it when implementing other providers, but hard to predict at this point